### PR TITLE
Math: Added epsilon param to .equals() of Vector2/3/4

### PR DESCRIPTION
**Pull request by Mugen87**
Solves #12165

This PR might be useful in other cases where strict equality checks lead to bad results because of floating point inaccuracies. PR is backward compatible and does not affect calls of `.equals()` without epsilon parameter.

### CODE REVIEW

File: src/extras/core/PathPrototype.js [hunk 1]
```diff
@@ -109,7 +109,7 @@ var PathPrototype = Object.assign( Object.create( CurvePath.prototype ), {
 			// if a previous curve is present, attempt to join
 			var firstPoint = curve.getPoint( 0 );
 
-			if ( ! firstPoint.equals( this.currentPoint ) ) {
+			if ( ! firstPoint.equals( this.currentPoint, 0.00001 ) ) {
```
Comment (Mugen87) [inline src/extras/core/PathPrototype.js]: `Number.EPSILON` is too small this case. A value like `0.00001` should do it...

File: src/math/Vector3.js [hunk 2]
```diff
@@ -683,9 +683,21 @@ Object.assign( Vector3.prototype, {
...
 
-		return ( ( v.x === this.x ) && ( v.y === this.y ) && ( v.z === this.z ) );
+		if ( epsilon === undefined ) {
+
+			return ( ( v.x === this.x ) && ( v.y === this.y ) && ( v.z === this.z ) );
+
+		} else {
+
+			var x = Math.abs( this.x - v.x );
+			var y = Math.abs( this.y - v.y );
```
Comment (WestLangley) [inline src/math/Vector3.js]: There is no need to compute `y` if the test for `x` fails.

**Comment by WestLangley** [comment]
See #7346 and #9170.

File: src/math/Vector3.js (hunk 2 above)
Comment (Mugen87) [reply src/math/Vector3.js]: So this pattern?
```javascript
var x = Math.abs( this.x - v.x );
if ( x > epsilon ) return false;

var y = Math.abs( this.y - v.y );
if ( y > epsilon ) return false;

return true;
```

**Comment by Mugen87** [comment]
Interesting. But at least we have with #12165 a valid use case where this enhancement makes sense.

File: src/math/Vector3.js (hunk 2 above)
Comment (WestLangley) [reply src/math/Vector3.js]: No. See #7346.

Comment (Mugen87) [reply src/math/Vector3.js]: Ah 😊 . I'll make a change...

Comment (Mugen87) [reply src/math/Vector3.js]: Love it like this!

**Comment by WestLangley** [comment]
I still think what I thought [previously](https://github.com/mrdoob/three.js/issues/7346#issuecomment-224470298).

But if you insist, I would call the second arg `tolerance`,  not `epsilon`, since depending on the scale of the values, the tolerance may not be small.

And I wouldn't name the method `equals`. I don't think we should be redefining the concept of equality.

You could call it `closeTo()`, but users may not like the distance metric you have selected.

**Comment by WestLangley** [comment]
Optionally, users can do this:
```javascript
vector.distanceToManhattan( vector2 ) < tolerance
```
BTW, we should rename that ridiculous name to `manhattanDistanceTo()`.

**Comment by Mugen87** [comment]
> BTW, we should rename that ridiculous name to manhattanDistanceTo().

:laughing::laughing::laughing:

**Comment by Mugen87** [comment]
With these new background information, i think it's better to close the issue and solve the mentioned issue with the manhatten distance.

**Pull request closed without merging**